### PR TITLE
Speed Up Whisper Export

### PR DESCRIPTION
**Pull request by petermcaughan**
### Description
Add a greedy option to the initializer deduplication process in the Whisper export.

Currently to detect shared initializers, ORT compares every initializer against every other initializer (n^2). In the comparison operator, if the two initializers have different data types (e.g. raw_data and int_64), both initializers are converted to a numpy array and the cast result is compared. This cast happens in every comparison, and exponentially affects the runtime of finding shared initializers. This cast operation is the bottleneck for the current Whisper export script.

The conversion to the numpy array is useful for detecting equal initializer values across nodes of different data types (e.g. recognizing a bias value of 0.0 is the same as a slice index of 0) but isn't triggered when comparing initializers of the same data type (e.g. weight value of 0.6 == weight value of 0.6). The latter case is where the majority of utility is for Whisper, and so by eliminating our path for comparing numpy arrays for initializers we save a lot of time for minimal cost.

In other words, this PR adds an option to remove the ability to detect shared initializers of different types (e.g. Slice Index and MatMul Constant) while retaining the ability to deduplicate weights.

### Motivation and Context
<!-- - Why is this change required? What problem does it solve?
- Current time to export Whisper-large is prohibitive.

### CODE REVIEW

**Comment by tianleiwu** [comment]
Instead of using such a tricky way, I suggest just add a threshold of initializer size (like 128 bytes) in comparison. You can follow the onnx code of using a threshold in saving external data format.

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 1]
```diff
@@ -1107,9 +1107,12 @@ def has_same_value(tensor1: TensorProto, tensor2: TensorProto) -> bool:
             return False
         if tensor1.HasField("raw_data") and tensor2.HasField("raw_data"):
             return tensor1.raw_data == tensor2.raw_data
+        if greedy:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model.py:1110]: rename greedy to require_raw_data

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 2]
```diff
@@ -1092,7 +1092,7 @@ def get_operator_statistics(self, include_domain=False):
         return op_count
 
     @staticmethod
-    def has_same_value(tensor1: TensorProto, tensor2: TensorProto) -> bool:
+    def has_same_value(tensor1: TensorProto, tensor2: TensorProto, greedy=False) -> bool:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model.py:1095]: update function document for the new parameter.

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 3]
```diff
@@ -1107,9 +1107,12 @@ def has_same_value(tensor1: TensorProto, tensor2: TensorProto) -> bool:
             return False
         if tensor1.HasField("raw_data") and tensor2.HasField("raw_data"):
             return tensor1.raw_data == tensor2.raw_data
+        if greedy:
+            return False
+
         return (numpy_helper.to_array(tensor1) == numpy_helper.to_array(tensor2)).all()
 
-    def remove_duplicated_initializer(self):
+    def remove_duplicated_initializer(self, greedy=False):
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model.py:1115]: update function document for the new parameter.

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 4]
```diff
@@ -889,6 +889,7 @@ def remove_shared_initializers(
     graph2: GraphProto,
     shared_prefix: str = "shared_",
     min_elements: int = 1024,
+    require_raw_data=False,
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/convert_generation.py:892]: Can you add type hints to the functions? Ex: `require_raw_data: bool = False`

File: onnxruntime/python/tools/transformers/models/whisper/whisper_chain.py [hunk 5]
```diff
@@ -141,7 +141,7 @@ def chain_model(args):
 
     # Initializers/opsets
     # Delete shared data between decoder/encoder and move to larger graph initializers
-    initializers = get_shared_initializers(encoder_model, decoder_model)
+    initializers = get_shared_initializers(decoder_model, encoder_model, require_raw_data=True)
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_chain.py:144]: I think the parameter order should be `(encoder_model, decoder_model, require_raw_data=True)` to match the function definition and to add the correct prefix to the names.
https://github.com/microsoft/onnxruntime/blob/85ad589c6f088a66b80ec864f59db4993a0b3ab8/onnxruntime/python/tools/transformers/convert_generation.py#L990-L998

Comment (petermcaughan) [reply onnxruntime/python/tools/transformers/models/whisper/whisper_chain.py:144]: This is true - and good observation with the prefixes. The reason I switched these around is because of how the loops are written in the deduplication process:

for remove_shared_initializers(graph A, graph B), the method will loop through every initializer in graph A, and see if they exist in graph B.

The decoder graph has most initializers shared with the encoderdecoderinit, while the encoderdecoderinit has fewer (half?) of initializers shared with encoderdecoderinit. 

From some basic testing, setting the decoder as graph A causes the loop to run fewer times (fewer initializers? randomness?) and caused the deduplication process to run about 10% faster. You're still right that the prefixes won't match up, but I think the speedup justifies the strangeness for the moment.

(I also think rewriting the code path to fix the prefixes in this case is a bit out of scope for this PR)

**Review by kunal-vaishnavi — APPROVED** [review]
